src/hal/src/buffer.rs

```diff
@@ -1,72 +1,58 @@
-//! Memory buffers
+//! Memory buffers.
+//!
+//! # Buffer
+//!
+//! Buffers interpret memory slices as linear continguous data array.
+//! They can be used as shader resources, vertex buffers, index buffers or for
+//! specifying the action commands for indirect exection.
 
-use std::error::Error;
-use std::fmt;
-
-use {IndexType, Backend};
+use {format, IndexType, Backend};
 
 
 /// An offset inside a buffer, in bytes.
 pub type Offset = u64;
 
-/// Error creating a buffer.
-#[derive(Clone, Copy, Debug, Eq, Hash, PartialEq)]
-pub enum CreationError {
-    /// Required `Usage` is not supported.
-    Usage(Usage),
-    /// Some other problem.
-    Other,
-}
-
-impl fmt::Display for CreationError {
-    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
-        let description = self.description();
-        match *self {
-            CreationError::Usage(usage) => write!(f, "{}: {:?}", description, usage),
-            _ => write!(f, "{}", description)
-        }
-    }
-}
-
-impl Error for CreationError {
-    fn description(&self) -> &str {
-        match *self {
-            CreationError::Usage(_) =>
-                "Required `Usage` is not supported",
-            CreationError::Other =>
-                "Some other problem",
-        }
-    }
-}
-
-/// Error creating a `BufferView`.
-#[derive(Clone, Debug, PartialEq)]
-pub enum ViewError {
-    /// The required usage flag is not present in the image.
-    Usage(Usage),
-    /// The backend refused for some reason.
-    Unsupported,
-}
+/// Buffer state.
+pub type State = Access;
 
-impl fmt::Display for ViewError {
-    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
-        let description = self.description();
-        match *self {
-            ViewError::Usage(usage) => write!(f, "{}: {:?}", description, usage),
-            _ => write!(f, "{}", description)
-        }
-    }
+/// Error creating a buffer.
+#[derive(Fail, Debug, Clone, PartialEq, Eq)]
+pub enum BufferCreationError {
```

Comment:
what's the motivation behind having the `Buffer*` prefix for those errors? If we are exposing `buffer` as a module, I find it elegant to keep it `buffer::CreationError`. If we are using the fully qualified names, there is no need to expose the modules.

Reply:
I'm not sure which direction we should go for the errors.
Currently having a `error` module containing `DeviceCreationError` and a few more. On the other side we could also move these errors into the different modules as done here with possible name shortening.

Reply:
I'm (non-strongly) in favor of having short error names in respective modules (as opposed to putting them into a dedicated `error` module)

Reply:
yeah I think this will be more clear. changed.